### ISSUE TITLE
Enhancement: [related-getter-setter-pairs] Option to allow undefined as type derivation for getters

### ISSUE THREAD

**Issue by tobbexiv** [typescript-eslint/typescript-eslint#10391]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/related-getter-setter-pairs/

### Description

The `https://typescript-eslint.io/rules/related-getter-setter-pairs/` is a great step to enforce consistent typing for getters and setters since typescript removed their type checking. However, one scenario which we have in our project and which might be common also for other projects is now reported: A combination of a setter with type `T` and getter with `T | undefined`. This combination allows for an unknown/undefined status before the value is set the first time, but prevents that after setting a value returning to this undefined status is possible. This example is also listed in the corresponding [Typescript release notes](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-1.html#unrelated-types-for-getters-and-setters).

I therefor suggest to extend the rule with an option that allows type `undefined` in addition to the setter types in the getter type definition.

### Fail

```TypeScript
Without the option: As current
With the option: All current except the newly excluded one
```

### Pass

```TypeScript
class A<T> {
  #val?: T

  set val(newVal: T) {
    this.#val = newVal;
  }

  get val(): T | undefined {
    return this.#val;
  }
}
```

**Comment by JoshuaKGoldberg** [comment]
Hmm, I can see why you'd ask for this, but it does feel odd to allow folks to write code where the setter can't take in the type of the getter. [[`a.val = a.val` not allowed in that example](https://typescript-eslint.io/play/#ts=5.6.2&fileType=.ts&code=MYGwhgzhAECCA8AVAfNA3gKGtAxANzBAH4AuaRDLaCAUwBdoCQAKAOxoHcA1QsxASnRVsdABYBLCADp8haAF5o7boQDcVAL6VsAc3qNCzfn2gAfaAFdWAExoAzce2tDs2AE70Lb1tDGSZTOrYWloYtqBgHtDAAPasEAxgZAgJbo46ANoAusjqGGBSTArQBYFAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aReAQ30QBNaAc0T5e0WshFjaxLpWioMkRNGgB7aJHBgAviB1A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false)]

What are the real world use cases for this? Why would you actually want the getter to have a wider type than the setter?

I also wonder if this should be a generalized option to allow a wider getter than setter - not specific to `| undefined`...

**Comment by tobbexiv** [comment]
All my real world use cases only have undefined in addition to the real type. I use it to reflect that the value to retrieve is not yet initialized or does not match certain expectations.

I currently have two use cases:
1. Optional ids which might not be defined on the object. But, if an id is to be defined, I don't want to allow setting it to undefined again.
2. DAO which allows to generically select properties (of objects not owned by me) and assert their type. In this scenario as well I don't want to allow setting invalid values. But, in case no valid value can be retrieved, this should reflect in the return value being undefined (+ logging in the selection function).

**Comment by JoshuaKGoldberg** [comment]
> DAO

What is DAO?

**Comment by christowiz** [comment]
The codebase I work on also an `| undefined` union.  This means that all these getters have to have a rule ignore comment.  The use cases are varied but in general if a class/SDK/dependency property is not set during the app lifecycle then an undefined return will trigger logic to "set" the value.

**Comment by kirkwaiblinger** [comment]
Just curious, why wouldn't you just want to have an explicit setter method for these kinds of cases?

```ts
class Foo {
}
```

**Comment by JoshuaKGoldberg** [comment]
It's been >3 months since the `evaluating community engagement` label was added on this issue and there's been no activity. We're going to go ahead and take this as a signal of non-interest. As in: this might be a useful idea, but there's not enough community interest to justify taking bandwidth away from other typescript-eslint initiatives to work on it or maintain it long-term. Closing out per https://typescript-eslint.io/maintenance/issues/#community-engagement-evaluation.

If someone feels strongly that this enhancement should exist, you can always fork the rule and write your own plugin. If your plugin then shows itself to be popular/useful we can always reconsider adding it to core.

Thanks for the discussion everyone! ❤
